SDWebImage/Core/SDGraphicsImageRenderer.m

```diff
@@ -0,0 +1,241 @@
+*
+* For the full copyright and license information, please view the LICENSE
+* file that was distributed with this source code.
+*/
+
+#import "SDGraphicsImageRenderer.h"
+#import "SDImageGraphics.h"
+
+@interface SDGraphicsImageRendererFormat ()
+#if SD_UIKIT
+@property (nonatomic, strong) UIGraphicsImageRendererFormat *uiformat API_AVAILABLE(ios(10.0), tvos(10.0));
+#endif
+@end
+
+@implementation SDGraphicsImageRendererFormat
+@synthesize scale = _scale;
+@synthesize opaque = _opaque;
+@synthesize preferredRange = _preferredRange;
+
+#pragma mark - Property
+- (CGFloat)scale {
+#if SD_UIKIT
+    if (@available(iOS 10.0, tvOS 10.10, *)) {
+        return self.uiformat.scale;
+    } else {
+        return _scale;
+    }
+#else
+    return _scale;
+#endif
+}
+
+- (void)setScale:(CGFloat)scale {
+#if SD_UIKIT
+    if (@available(iOS 10.0, tvOS 10.10, *)) {
+        self.uiformat.scale = scale;
+    } else {
+        _scale = scale;
+    }
+#else
+    _scale = scale;
+#endif
+}
+
+- (BOOL)opaque {
+#if SD_UIKIT
+    if (@available(iOS 10.0, tvOS 10.10, *)) {
+        return self.uiformat.opaque;
+    } else {
+        return _opaque;
+    }
+#else
+    return _opaque;
+#endif
+}
+
+- (void)setOpaque:(BOOL)opaque {
+#if SD_UIKIT
+    if (@available(iOS 10.0, tvOS 10.10, *)) {
+        self.uiformat.opaque = opaque;
+    } else {
+        _opaque = opaque;
+    }
+#else
+    _opaque = opaque;
+#endif
+}
+
+- (SDGraphicsImageRendererFormatRange)preferredRange {
+#if SD_UIKIT
+    if (@available(iOS 10.0, tvOS 10.10, *)) {
+        if (@available(iOS 12.0, tvOS 12.0, *)) {
+            return (SDGraphicsImageRendererFormatRange)self.uiformat.preferredRange;
+        } else {
+            BOOL prefersExtendedRange = self.uiformat.prefersExtendedRange;
+            if (prefersExtendedRange) {
+                return SDGraphicsImageRendererFormatRangeExtended;
+            } else {
+                return SDGraphicsImageRendererFormatRangeStandard;
+            }
+        }
+    } else {
+        return _preferredRange;
+    }
+#else
+    return _preferredRange;
+#endif
+}
+
+- (void)setPreferredRange:(SDGraphicsImageRendererFormatRange)preferredRange {
+#if SD_UIKIT
+    if (@available(iOS 10.0, tvOS 10.10, *)) {
+        if (@available(iOS 12.0, tvOS 12.0, *)) {
+            self.uiformat.preferredRange = (UIGraphicsImageRendererFormatRange)preferredRange;
+        } else {
+            switch (preferredRange) {
+                case SDGraphicsImageRendererFormatRangeExtended:
+                    self.uiformat.prefersExtendedRange = YES;
+                    break;
+                case SDGraphicsImageRendererFormatRangeStandard:
+                    self.uiformat.prefersExtendedRange = NO;
+                default:
+                    // Automatic means default
+                    break;
+            }
+        }
+    } else {
+        _preferredRange = preferredRange;
+    }
+#else
+    _preferredRange = preferredRange;
+#endif
```

Comment:
Maybe  `#if SD_UIKIT` is unnecessary.

Reply:
`UIGraphicsImageRendererFormatRange` this symbol does not exist on watchOS/macOS SDK. Even you use the `@available (iOS, *)`, this cause the compile failed. You can have a try.

Reply:
get !